File: server/whisper.py
"""Whisper as a general tool: transcribe any audio or video file, and time lyrics.

Usage: whisper.py <input> [--lyrics <lyrics.txt>] [--out <out-stem>]
                  [--language <code>] [--words] [--vocals <vocals.flac>]

Prints one JSON object on stdout, on success AND on every failure Python can
still report, the same contract as lrc.py (server/lrc.js reads both the same
way, and its CPU re-run after a native cuDNN abort works here unchanged).

What it answers:
  - always: language, text, segments [{start, end, text}], device, model;
  - --words: each segment carries words [{text, start, end}];
  - --lyrics: the known lyrics reconciled against what was heard, with lrc.py's
    reconcile() (our text, whisper's timing): lines [{start, text}] (and their
    words with --words), confidence, matched;
  - --out: <out-stem>.lrc and <out-stem>.word.lrc, from the known lyrics when
    given, else from the heard segments.

Environment: AIPLAY_WHISPER_MODEL and AIPLAY_WHISPER_DEVICE, exactly as lrc.py
reads them. The device logic, the failure sentences and the device marker on
stderr are lrc.py's own, imported rather than copied: a second copy is how two
scripts come to disagree about which GPU failure falls back to the CPU.
"""
from __future__ import annotations

import gc
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import lrc  # noqa: E402  (beside this file; stdlib-only at import)

USAGE = "whisper.py <input> [--lyrics <file>] [--out <stem>] [--language <code>] [--words] [--vocals <file>]"


def parse_args(argv: list[str]) -> dict:
    """argv -> options. Unknown flags are an error, not ignored: a caller that
    misspells --lyrics must hear about it rather than get a plain transcript."""
    if not argv or argv[0].startswith("--"):
        raise lrc.LrcError(f"no input file; usage: {USAGE}")
    opts = {"input": argv[0], "lyrics": None, "out": None, "language": None, "words": False, "vocals": None}
    i = 1
    while i < len(argv):
        a = argv[i]
        if a == "--words":
            opts["words"] = True
            i += 1
            continue
        if a in ("--lyrics", "--out", "--language", "--vocals"):
            if i + 1 >= len(argv):
                raise lrc.LrcError(f"{a} needs a value; usage: {USAGE}")
            opts[a[2:]] = argv[i + 1]
            i += 2
            continue
        raise lrc.LrcError(f"unknown argument {a!r}; usage: {USAGE}")
    lang = (opts["language"] or "").strip().lower()
    opts["language"] = None if lang in ("", "auto") else lang
    return opts


def transcribe_full(audio: str, device: str, compute_type: str, language: str | None) -> dict:
    """Segments with word timing, the language whisper settled on, the text.

    The same model call as lrc.transcribe() (VAD on, silence suppressed,
    regrouped, progress bars off stdout), keeping what that one throws away."""
    import stable_whisper
    lrc.PHASE = "load"
    model = stable_whisper.load_faster_whisper(lrc.MODEL, device=device, compute_type=compute_type)
    lrc.PHASE = "transcribe"
    kw = {"language": language} if language else {}
    res = model.transcribe(audio, word_timestamps=True, suppress_silence=True,
                           vad=True, regroup=True, verbose=None, **kw)
    segments = []
    for seg in res.segments:
        text = (getattr(seg, "text", "") or "").strip()
        words = []
        for w in (seg.words or []):
            t = (getattr(w, "word", "") or "").strip()
            if t and getattr(w, "start", None) is not None:
                words.append({"text": t, "start": round(float(w.start), 3),
                              "end": round(float(w.end if w.end is not None else w.start), 3)})
        if not text and not words:
            continue
        segments.append({"start": round(float(seg.start), 3), "end": round(float(seg.end), 3),
                         "text": text, "words": words})
    return {
        "language": getattr(res, "language", None) or language,
        "text": " ".join(s["text"] for s in segments).strip(),
        "segments": segments,
    }


def heard_words(segments: list[dict]) -> list[dict]:
    """The flat word list reconcile() takes as its clock."""
    return [w for s in segments for w in s["words"]]


def segment_lines(segments: list[dict]):
    """Heard segments in the (known, flat, times) shape lrc.build() writes, so a
    transcript with no known lyrics still makes an LRC pair. A segment whisper
    gave no word timing gets its own start for every word."""
    known, flat, times = [], [], []
    for s in segments:
        words = [(w["text"], w["start"]) for w in s["words"]] or [(t, s["start"]) for t in s["text"].split()]
        if not words:
            continue
        li = len(known)
        known.append([t for t, _ in words])
        for wi, (t, at) in enumerate(words):
            flat.append((li, wi, t))
            times.append(at)
    return known, flat, times


def aligned_lines(known, flat, times, with_words: bool) -> list[dict]:
    out: dict[int, dict] = {}
    for (li, _wi, w), t in zip(flat, times):
        line = out.setdefault(li, {"start": round(float(t), 3), "text": " ".join(known[li])})
        if with_words:
            line.setdefault("words", []).append({"text": w, "start": round(float(t), 3)})
    return [out[li] for li in range(len(known)) if li in out]


def run(opts: dict) -> dict:
    """The transcription, on the device lrc.pick_device() chose, moving to the
    CPU in this process after a GPU failure Python can catch: lrc.py's main()
    line for line, because the reasons are the same (see there)."""
    device = compute = why_cpu = None
    source = opts["vocals"] if opts["vocals"] and os.path.exists(opts["vocals"]) else None
    audio = source or opts["input"]
    got = None
    try:
        import stable_whisper  # noqa: F401  (maps a CUDA torch's libraries first)
        device, compute, why_cpu = lrc.pick_device()
        lrc.mark_device(device, compute)
        try:
            got = transcribe_full(audio, device, compute, opts["language"])
        except Exception as exc:  # noqa: BLE001
            if not (device == "cuda" and lrc.wanted_device() == "auto" and lrc.gpu_failure(exc)):
                raise
            why_cpu = f"the GPU run failed ({lrc.one_line(exc, 200)})"
        if got is None:
            gc.collect()
            device, compute = "cpu", "int8"
            lrc.mark_device(device, compute)
            got = transcribe_full(audio, device, compute, opts["language"])
    except Exception as exc:  # noqa: BLE001
        error = lrc.describe(exc, device)
        if device == "cpu" and why_cpu and why_cpu != "AIPLAY_WHISPER_DEVICE=cpu":
            error += f" (on the CPU because {why_cpu})"
        return {"ok": False, "error": error[:600], "device": device, "needsInstall": lrc.needs_install(exc)}
    lrc.PHASE = "write"

    segments = got["segments"]
    result = {
        "ok": True,
        "language": got["language"],
        "text": got["text"],
        "segments": [s if opts["words"] else {k: s[k] for k in ("start", "end", "text")} for s in segments],
        "duration": segments[-1]["end"] if segments else 0.0,
        "usedVocalStem": bool(source),
        "device": device,
        "computeType": compute,
        "model": lrc.MODEL,
    }
    if device == "cpu":
        result["cpuReason"] = why_cpu

    known = None
    if opts["lyrics"]:
        with open(opts["lyrics"], encoding="utf-8") as fh:
            known = lrc.lyric_lines(fh.read())
    if known:
        heard = heard_words(segments)
        flat, times, matched = lrc.reconcile(known, heard)
        result["aligned"] = {
            "lines": aligned_lines(known, flat, times, opts["words"]),
            "words": len(flat), "heard": len(heard), "matched": matched,
            # How much of the timing is measured rather than interpolated.
            "confidence": round(matched / max(1, len(flat)), 3),
        }
    elif opts["lyrics"]:
        result["alignNote"] = "the lyrics had no lines to time (only section markers, or empty)"
    else:
        known, flat, times = segment_lines(segments)

    if opts["out"]:
        if known:
            line_lrc, word_lrc = lrc.build(known, flat, times)
            with open(opts["out"] + ".lrc", "w", encoding="utf-8") as fh:
                fh.write(line_lrc)
            with open(opts["out"] + ".word.lrc", "w", encoding="utf-8") as fh:
                fh.write(word_lrc)
            result["lrc"] = os.path.basename(opts["out"]) + ".lrc"
            result["wordLrc"] = os.path.basename(opts["out"]) + ".word.lrc"
        else:
            result["lrcNote"] = "nothing was heard, so no LRC was written"
    return result


def main() -> int:
    opts = parse_args(sys.argv[1:])
    if not os.path.isfile(opts["input"]):
        raise lrc.LrcError(f"there is no file at {opts['input']}")
    lrc.emit(run(opts))
    return 0


if __name__ == "__main__":
    # Same net as lrc.py: anything Python can still catch leaves one JSON line.
    try:
        code = main()
    except SystemExit:
        raise
    except BaseException as exc:  # noqa: BLE001
        lrc.emit({"ok": False, "error": lrc.describe(exc)[:600], "needsInstall": lrc.needs_install(exc)})
        code = 0
    sys.exit(code)

File: server/whisper.js
/**
 * Whisper as a tool anything in Studio can call: POST/GET /api/whisper, and
 * through it the whisper_transcribe and whisper_status MCP tools.
 *
 * Timed lyrics were the only way in: a library song, its own lyrics, two LRC
 * files. An agent that needs the words of a clip, the lyrics of a song nobody
 * wrote down, or the timing of lyrics it was handed had no door. This is that
 * door, and it adds no second whisper:
 *
 *   - the SAME python (config.lyrics.python) and the same model
 *     (config.lyrics.model), so one Set up timed lyrics serves both;
 *   - the SAME runner (server/lrc.js runLrc over server/whisper.py, which
 *     imports lrc.py's device logic), so the failures read the same;
 *   - the SAME queue (art.js, kind "whisper", a program of its own like
 *     "lrc"): it waits for music, never runs beside a picture, clip or
 *     separation, and Stop kills its tree.
 *
 * WHAT IT MAY READ. A library song by file name, a clip or an imported file by
 * its name in the clips folder (/api/studio/import lands there), or a path
 * that resolves, symlinks followed, inside Studio's output folder. Nothing
 * else: a path is a way to name a stem or a render that has no library name,
 * never a way to read the rest of the disk. Every POST is same-origin local
 * JSON only (sameOriginLocalJson, the guard of the other doors that run a
 * program), so another web page cannot queue work or choose the model.
 */
import { realpath, stat } from "node:fs/promises";
import path from "node:path";
import { randomUUID } from "node:crypto";
import { WHISPER_MODELS } from "./config.js";
import { whisperPythonMissing, pythonVerdict, installLines } from "./lrc.js";
import { actorFrom } from "./provenance.js";

/** What whisper.py (ffmpeg underneath) is handed. Pictures are not. */
export const WHISPER_AUDIO = /\.(flac|mp3|opus|wav|ogg|m4a|aac)$/i;
export const WHISPER_MEDIA = /\.(flac|mp3|opus|wav|ogg|m4a|aac|mp4|webm|mov|mkv|m4v)$/i;
/** Whisper's language codes are two letters, three for a few ("haw", "yue"). */
const LANGUAGE_RE = /^[a-z]{2,3}$/;
export const MAX_LYRICS = 20_000;

const badName = (s) => !s || s.includes("..") || s.includes("/") || s.includes("\\") || /[\r\n\0]/.test(s);

/**
 * The body of {action:"transcribe"} → { spec } or { error }. Checks shape only;
 * resolveWhisperInput() checks the disk.
 */
export function parseWhisperRequest(b = {}) {
  const named = ["file", "clip", "path"].filter((k) => b[k] !== undefined && b[k] !== null && b[k] !== "");
  if (named.length !== 1) {
    return { error: "Name exactly one of file (a library song), clip (a clip or imported file) or path (a file inside Studio's output folder)." };
  }
  const by = named[0];
  const value = b[by];
  if (typeof value !== "string" || value.length > 1024) return { error: `${by} must be a name or path of up to 1024 characters.` };
  if (by !== "path" && badName(value)) return { error: `Bad ${by} name: ${JSON.stringify(value)}.` };
  if (b.lyrics !== undefined && b.lyrics !== null && typeof b.lyrics !== "string") return { error: "lyrics must be text." };
  const lyrics = String(b.lyrics || "").trim();
  if (lyrics.length > MAX_LYRICS) return { error: `lyrics are limited to ${MAX_LYRICS} characters.` };
  let language = b.language === undefined || b.language === null ? "" : String(b.language).trim().toLowerCase();
  if (language === "auto") language = "";
  if (language && !LANGUAGE_RE.test(language)) return { error: "language must be a whisper language code such as en, de or ja, or auto." };
  for (const k of ["words", "writeLrc", "vocals"]) {
    if (b[k] !== undefined && b[k] !== null && typeof b[k] !== "boolean") return { error: `${k} must be true or false.` };
  }
  return {
    spec: {
      by, value, lyrics: lyrics || null, language: language || null,
      words: b.words === true, writeLrc: b.writeLrc === true,
      /* null: the timed-lyrics setting decides (config.lyrics.useVocalStem). */
      vocals: typeof b.vocals === "boolean" ? b.vocals : null,
    },
  };
}

/** Is `child` inside `root` (both already real paths)? */
export function insideFolder(root, child, p = path) {
  const rel = p.relative(root, child);
  return !!rel && !rel.startsWith("..") && !p.isAbsolute(rel);
}

async function isFile(f) {
  try { return (await stat(f)).isFile(); } catch { return false; }
}

/**
 * The spec → the file whisper.py reads, a stem for its LRC files, the vocal
 * stem when one is used, and a title. → { input, name, lrcBase, vocals, title }
 * or { error, status }. `dirs` = { outputDir, clipDir, stemsModel, useVocalStem }.
 */
export async function resolveWhisperInput(spec, dirs) {
  const { outputDir, clipDir, stemsModel = "htdemucs_ft", useVocalStem = true } = dirs;
  let input, name, lrcBase, vocals = null;
  if (spec.by === "file") {
    if (!WHISPER_AUDIO.test(spec.value)) return { error: `${spec.value} is not a library song (flac, mp3, opus or wav).`, status: 400 };
    input = path.join(outputDir, spec.value);
    name = spec.value;
    lrcBase = spec.value.replace(WHISPER_MEDIA, "");
    /* A clean vocal transcribes better than a mix, the reason timed lyrics use
     * it: the song's separated vocal when one exists, never separated for this. */
    const want = spec.vocals ?? useVocalStem;
    const stem = path.join(outputDir, "stems", stemsModel, lrcBase, "vocals.flac");
    if (want && await isFile(stem)) vocals = stem;
  } else if (spec.by === "clip") {
    if (!WHISPER_MEDIA.test(spec.value)) return { error: `${spec.value} is not audio or video.`, status: 400 };
    input = path.join(clipDir, spec.value);
    name = spec.value;
    lrcBase = spec.value.replace(WHISPER_MEDIA, "");
  } else {
    const raw = spec.value.trim().replace(/^"(.*)"$/, "$1").trim();
    /* \\server\share and \\?\ are absolute to path.win32: another machine's
     * file, or a device. Only this disk, only Studio's own folder. */
    if (/^[\\/]{2}/.test(raw) || !path.isAbsolute(raw) || /[\r\n\0]/.test(raw)) {
      return { error: "path must be the full path of a file inside Studio's output folder.", status: 400 };
    }
    if (!WHISPER_MEDIA.test(raw)) return { error: `${path.basename(raw)} is not audio or video.`, status: 400 };
    let real, root;
    try { real = await realpath(raw); } catch { return { error: `There is no file at ${raw}.`, status: 404 }; }
    try { root = await realpath(outputDir); } catch { root = path.resolve(outputDir); }
    if (!insideFolder(root, real)) {
      return { error: `Only files inside Studio's output folder (${root}) can be transcribed by path.`, status: 403 };
    }
    input = real;
    name = path.basename(real);
    /* The folders it sits in are part of its name here: every separated song
     * has a vocals.flac, and they must not write over each other's LRC. */
    lrcBase = path.relative(root, real).replace(WHISPER_MEDIA, "").replace(/[\\/]+/g, "_");
  }
  if (!await isFile(input)) return { error: `There is no file ${name}.`, status: 404 };
  /* Characters no Windows file name may hold become "_"; letters in any
   * script stay, as they do in the song's own file name. */
  return { input, name, lrcBase: lrcBase.replace(/[<>:"/\\|?*\x00-\x1f]+/g, "_").slice(0, 160) || "whisper", vocals, title: `Transcribe ${name}` };
}

/** Where the runner's job for this id is: running, queued, finished or gone. */
export function whisperJob(art, id) {
  const mine = (j) => j && j.id === id && j.kind === "whisper";
  if (mine(art.current)) return { state: "running", job: art.current };
  const at = (art.queue || []).findIndex(mine);
  if (at >= 0) return { state: "queued", position: at + 1, job: art.queue[at] };
  const done = (art.done || []).find(mine);
  if (!done) return { state: "missing", job: null };
  if (done.error) return { state: done.cancelled ? "stopped" : "failed", job: done };
  return { state: "done", job: done };
}

/** A job as the door reports it: never the internals, the result when there is one. */
export function jobAnswer(id, found) {
  const j = found.job;
  return {
    id, state: found.state,
    ...(found.position ? { position: found.position } : {}),
    ...(j ? { title: j.title || null } : {}),
    ...(j?.error ? { error: String(j.error) } : {}),
    ...(found.state === "done" ? { result: j.transcript || null } : {}),
  };
}

/**
 * The /api/whisper door. index.js mounts it the way it mounts the prompt tools.
 * `probe(py, modules)` is index.js's probeOne; `modules` its LYRICS_MODULES.
 */
export function createWhisperRoutes({ json, readBody, sameOriginLocalJson, art, config, probe, modules, savePrefs, lrcDir, clipDir }) {
  async function status() {
    const py = config.lyrics.python;
    const got = whisperPythonMissing(py) ? {} : await probe(py, modules).catch(() => ({}));
    const v = pythonVerdict({ python: py, chosen: config.lyrics.whisperPython, modules: Object.fromEntries(modules.map((m) => [m, !!got[m]])) });
    const mine = [art.current, ...(art.queue || [])].filter((j) => j?.kind === "whisper");
    return {
      ...v,
      ...(v.ready ? {} : { install: installLines(py), ...(process.env.AIPLAY_WHISPER_PYTHON ? {} : { setup: "lyrics" }) }),
      model: config.lyrics.model,
      models: WHISPER_MODELS,
      device: String(process.env.AIPLAY_WHISPER_DEVICE || "auto").trim().toLowerCase(),
      jobs: mine.map((j) => ({ id: j.id, title: j.title, state: j === art.current ? "running" : "queued" })),
    };
  }

  return async function routes(req, res, url) {
    const p = url.pathname;
    if (p !== "/api/whisper") return false;
    if (req.method === "GET") {
      const id = url.searchParams.get("job");
      if (id) {
        const found = whisperJob(art, id);
        if (found.state === "missing") {
          return json(res, 404, { error: `No whisper job ${id}: not running, not queued and not among the finished jobs (Studio may have restarted).` }), true;
        }
        return json(res, 200, { ok: true, job: jobAnswer(id, found) }), true;
      }
      return json(res, 200, { ok: true, whisper: await status() }), true;
    }
    if (req.method !== "POST") return json(res, 405, { error: "GET or POST." }), true;
    /* Every action runs a program or changes a setting: Studio's page and
     * local clients (MCP) only, the same guard as choosing the python. */
    if (!sameOriginLocalJson(req)) {
      return json(res, 403, { error: "Whisper requires a same-origin local JSON request." }), true;
    }
    const b = (await readBody(req)) || {};
    if (b.action === "status") return json(res, 200, { ok: true, whisper: await status() }), true;
    if (b.action === "model") {
      if (b.value !== undefined) {
        if (!WHISPER_MODELS.includes(b.value)) {
          return json(res, 400, { error: `The whisper model must be one of ${WHISPER_MODELS.join(", ")}.` }), true;
        }
        /* Read at every spawn (art.js), so the next job uses it: no restart. */
        config.lyrics.model = b.value;
        await savePrefs();
      }
      return json(res, 200, { ok: true, whisper: await status() }), true;
    }
    if (b.action === "transcribe") {
      const parsed = parseWhisperRequest(b);
      if (parsed.error) return json(res, 400, { error: parsed.error }), true;
      const spec = parsed.spec;
      const at = await resolveWhisperInput(spec, {
        outputDir: config.outputDir, clipDir,
        stemsModel: config.stems?.model, useVocalStem: config.lyrics.useVocalStem !== false,
      });
      if (at.error) return json(res, at.status || 400, { error: at.error }), true;
      /* Refused here, where the caller can act on it, rather than queued to
       * die: the same sentence and setup id as Time the lyrics. */
      const noPython = whisperPythonMissing(config.lyrics.python);
      if (noPython) return json(res, 400, { error: noPython, ...(process.env.AIPLAY_WHISPER_PYTHON ? {} : { setup: "lyrics" }) }), true;
      const job = art.request({
        file: `whisper:${randomUUID().slice(0, 8)}`, title: at.title, kind: "whisper", asked: true,
        actor: actorFrom(req),
        whisper: {
          input: at.input, vocals: at.vocals, lyrics: spec.lyrics, language: spec.language, words: spec.words,
          /* ".whisper" keeps these apart from the song's own timed lyrics
           * (<song>.lrc), which a transcription must never overwrite. */
          outStem: spec.writeLrc ? path.join(lrcDir, `${at.lrcBase}.whisper`) : null,
        },
      });
      if (!job) return json(res, 409, { error: `Not queued: ${art.lastRefusal || "the queue refused it"}.` }), true;
      return json(res, 200, {
        ok: true, jobId: job.id, job: { id: job.id, kind: job.kind, title: job.title },
        input: at.name, usesVocalStem: !!at.vocals, model: config.lyrics.model,
        ...(spec.writeLrc ? { lrc: `${at.lrcBase}.whisper.lrc`, wordLrc: `${at.lrcBase}.whisper.word.lrc` } : {}),
      }), true;
    }
    return json(res, 400, { error: "Unknown action. Try: transcribe, model, status." }), true;
  };
}
